src/platform/android/java/chip/platform/AndroidBleManager.java

```diff
@@ -602,14 +609,23 @@ public void onConnectionStateChange(BluetoothGatt gatt, int status, int newState
     public void onServicesDiscovered(BluetoothGatt gatt, int status) {
       Log.d(TAG, "onServicesDiscovered status = " + status);
       super.onServicesDiscovered(gatt, status);
+      if (mState != STATE_DISCOVER_SERVICE) {
+        Log.d(TAG, "Invalid state : " + mState);
+        return;
+      }
 
       Log.i(TAG, "Services Discovered");
+      mState = STATE_REQUEST_MTU;
       gatt.requestMtu(247);
```

Comment:
The request of MTU from here is probably what broke on the devices. gatt calls should not be done inside of AndroidBluetoothGattCallback. See https://developer.android.com/develop/connectivity/bluetooth/ble/connect-gatt-server#broadcast-updates for guidance, where it is required to post to a different context to properly handle callbacks rather than calling back in, which can be unsafe on some Android platforms.

Reply:
The previous state of the Android driver had the same problem, but adding complex statefulness to this file may add risk and hard-to-debug issues in other situations.